### PR TITLE
[SPARK-7829][CORE][WIP] multiple attempts of shuffle tasks is OK

### PR DESCRIPTION
**Pull request by squito**
fix for one of the retry issues -- data files are appended to, but the index file always points to the beginning of the file.  This would lead to an apparently corrupted data file during the shuffle fetch.

for now this is just for discussion, not meant to merge until we flesh out some other options more.  Even if in master we make larger changes to deal w/ stage retry, this is small enough to potentially be backported to earlier branches (maybe without some of the hokey test code).

cc @JoshRosen

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/shuffle/unsafe/UnsafeShuffleSuite.scala [hunk 1]
```diff
@@ -41,6 +41,20 @@ class UnsafeShuffleSuite extends ShuffleSuite with BeforeAndAfterAll {
...
+    val kryoAndNoSpillMove = conf.clone()
+      .set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
+      .set("spark.shuffle.unsafe.testing.allowSpillMove", "false")
+    val noCompression = kryoAndNoSpillMove.clone()
+      .set("spark.shuffle.compress", "false")
+      .set("spark.file.transferTo","false")
+    Seq(
+      "slow merge path" -> kryoAndNoSpillMove,
+      "filestream based fast merge" -> noCompression
+    )
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/shuffle/unsafe/UnsafeShuffleSuite.scala]: there should be another case here for the `transferTo` path as well, but that path fails with the msg about "Please check your kernel version to see if it is 2.6.32".  I don't understand the intricacies of when the channel's position is wrong from the discussion here https://issues.apache.org/jira/browse/SPARK-3948 so I just left it as is.

(which is another reason to consider an approach where the same files are written to, so append vs. overwrite doesn't matter)

Comment (JoshRosen) [inline core/src/test/scala/org/apache/spark/shuffle/unsafe/UnsafeShuffleSuite.scala]: Did it fail when running the test locally?  Which Linux kernel version are you running?

Comment (squito) [inline core/src/test/scala/org/apache/spark/shuffle/unsafe/UnsafeShuffleSuite.scala]: oh no, sorry I wasn't clear.  It did fail when running locally, but only b/c of the extra check we have in place [here](https://github.com/apache/spark/blob/b463e6d618e69c535297e51f41eca4f91bd33cc8/core/src/main/java/org/apache/spark/shuffle/unsafe/UnsafeShuffleWriter.java#L390).

What I meant was, I could change that check, to take into account the size of the file initially, but I wasn't sure if that would be reliable on all systems given SPARK-3948.

**Comment by squito** [comment]
closing this in favor of https://github.com/apache/spark/pull/6648

**Pull request closed without merging**